testcrate/ui/bad-return-type.stderr

```diff
@@ -10,5 +10,29 @@ error[E0308]: mismatched types
    = note: expected type `std::option::Option<i32>`
               found type `{integer}`
 
-error: aborting due to previous error
+error[E0308]: mismatched types
+  --> $DIR/bad-return-type.rs:17:36
+   |
+17 |    #[async_stream(item = Option<i32>)]
+   |   ____________________________________^
```

Comment:
This looks like a particularly confusing error message, especially with respect to span information. Is there anything we can do to improve this?

Reply:
The horribleness of this error message comes from `stream_yield!` expanding to

```rust
let e = $e;
yield ::futures::Async::Ready(e)
```

the span which is in error is the `e` in the second line of the expansion (although, why it then highlights what it does I don't know). Changing it to simply 

```rust
yield ::futures::Async::Ready($e)
```

changes the error message to what it should be:

```
error[E0308]: mismatched types
 --> $DIR/bad-return-type.rs:25:19
   |
25 |     stream_yield!(val);
   |                   ^^^
   |                   |
   |                   expected enum `std::option::Option`, found integral variable
   |                   help: try using a variant of the expected type: `Some(val)`
   |
   = note: expected type `std::option::Option<_>`
              found type `{integer}`
```

but causes test failures

```
error[E0597]: `t` does not live long enough
   --> tests/smoke.rs:98:19
    |
98  |     stream_yield!(t.clone());
    |                   ^ does not live long enough
...
101 | }
    | - borrowed value only lives until here
    |
    = note: borrowed value must be valid for the static lifetime...
```

I believe those are related to https://github.com/rust-lang/rust/issues/44197, once that is resolved then the `stream_yield!` expansion can be updated and this should be fixed.

I'd prefer to merge with the current error message and I'll open an issue to fix it once the upstream issue is fixed.

Reply:
The alternative is to change the expansion now and change the test to

```
let c = t.clone();
stream_yield!(c);
```

Doing this instead would increase the churn once rust-lang/rust#44197 is fixed, every user of this library that had to make an extra temporary would have to fix their code at some point. As long as that issue doesn't take too long to fix I think we can live with the worse error message for a short while.

Reply:
Ok yeah this sounds like a fine tradeoff to me, mind leaving a comment in the definition of the `stream_yield!` macro pointing to the upstream issue?